### DISCUSSION TITLE
Environment Details Query

### DISCUSSION THREAD

**Discussion by ghost** [d365collaborative/d365fo.tools#533] [Ideas]
One of the more common tasks I need to perform is to update a worksheet with details about an environment, such as "last package applied" and "last database refresh" which are items listed in the History of the environment. Is it possible to query those details through the LCS API? It would also be helpful to query details like the URL for the environment, the service update version, and other common details that I need to keep track of when managing 30+ environments on our project.

**Comment by Splaxi** [comment]
Hi @aaronstackpole 

Thanks for reaching out, but I'm sorry to be the bringer of bad news. There isn't any endpoints in the LCS API, with these details - yet.

Unofficial rumors has it that a single engineer within the LCS team at Microsoft has heard our wishes and has taken up the task, outside the ordinary road map, and will be releasing something in the near future. 

@richardsondev should be able to share more information.

**Reply by richardsondev** [reply]
Thanks @Splaxi for the ping.

I am happy to announce that we have three new LCS OAuth endpoints that are being reviewed internally. 

1. Fetch an environment's information (current status, environment url, and other metadata)
2. Fetch change history entries for an environment (with asset id, time completed, status, etc)
3. Fetch all environments in a given project

We have these already working but need to make sure they have the polish required for release. I will keep you all posted and hopefully will have a PR into this codebase soon with the functions to access these endpoints.

Regards,
Billy

**Reply by ghost** [reply]
This is fantastic!

**Reply by richardsondev** [reply]
Hi @aaronstackpole , to loop back on this discussion: LCS has since released the above APIs and D365FO.Tools supports them.

Please find the documentation below. Points 1 and 3 are combined within a single endpoint since my initial message here.
https://docs.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/lifecycle-services/api/v1/reference-environment-metadata
https://docs.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/lifecycle-services/api/v1/reference-environment-history

D365FO.tools docs: 
https://github.com/d365collaborative/d365fo.tools/blob/development/docs/Get-D365LcsEnvironmentMetadata.md
https://github.com/d365collaborative/d365fo.tools/blob/development/docs/Get-D365LcsEnvironmentHistory.md

Regards,
Billy
